lib/fileio/commands/baseExportCommand.cpp

```diff
@@ -0,0 +1,316 @@
+}
+
+/* virtual */
+std::unique_ptr<UsdMaya_WriteJob> MayaUSDExportCommand::initializeWriteJob(const pxr::UsdMayaJobExportArgs & args)
+{
+    return std::unique_ptr<UsdMaya_WriteJob>(new UsdMaya_WriteJob(args));
+}
+
+MStatus MayaUSDExportCommand::doIt(const MArgList & args)
+{
+try
+{
+    MStatus status;
+
+    MArgDatabase argData(syntax(), args, &status);
+
+    // Check that all flags were valid
+    if (status != MS::kSuccess) {
+        return status;
+    }
+
+    // Read all of the dictionary args first.
+    const VtDictionary userArgs = UsdMayaUtil::GetDictionaryFromArgDatabase(
+            argData, UsdMayaJobExportArgs::GetDefaultDictionary());
+
+    // Now read all of the other args that are specific to this command.
+    bool append = false;
+    std::string fileName;
+
+    if (argData.isFlagSet(kAppendFlag)) {
+        argData.getFlagArgument(kAppendFlag, 0, append);
+    }
+
+    if (argData.isFlagSet(kFileFlag))
+    {
+        // Get the value
+        MString tmpVal;
+        argData.getFlagArgument(kFileFlag, 0, tmpVal);
+
+        // resolve the path into an absolute path
+        MFileObject absoluteFile;
+        absoluteFile.setRawFullName(tmpVal);
+        absoluteFile.setRawFullName( absoluteFile.resolvedFullName() ); // Make sure an absolute path
+        fileName = absoluteFile.resolvedFullName().asChar();
+
+        if (fileName.empty()) {
+            fileName = tmpVal.asChar();
+        }
+    }
+    else {
+        TF_RUNTIME_ERROR("-file not specified.");
+        return MS::kFailure;
+    }
+
+    if (fileName.empty()) {
+        return MS::kFailure;
+    }
+
+    // If you provide a frame range we consider this an anim
+    // export even if start and end are the same
+    GfInterval timeInterval;
+    if (argData.isFlagSet(kFrameRangeFlag)) {
+        double startTime = 1;
+        double endTime = 1;
+        argData.getFlagArgument(kFrameRangeFlag, 0, startTime);
+        argData.getFlagArgument(kFrameRangeFlag, 1, endTime);
+        if (startTime > endTime) {
+            // If the user accidentally set start > end, resync to the closed
+            // interval with the single start point.
+            timeInterval = GfInterval(startTime);
+        }
+        else {
+            // Use the user's interval as-is.
+            timeInterval = GfInterval(startTime, endTime);
+        }
+    } else {
+        // No animation, so empty interval.
+        timeInterval = GfInterval();
+    }
+
+    double frameStride = 1.0;
+    if (argData.isFlagSet(kFrameStrideFlag)) {
+        argData.getFlagArgument(kFrameStrideFlag, 0, frameStride);
+    }
+
+    std::set<double> frameSamples;
+    unsigned int numFrameSamples = argData.numberOfFlagUses(kFrameSampleFlag);
+    for (unsigned int i = 0; i < numFrameSamples; ++i) {
+        MArgList tmpArgList;
+        argData.getFlagArgumentList(kFrameSampleFlag, i, tmpArgList);
+        frameSamples.insert(tmpArgList.asDouble(0));
+    }
```

Comment:
This is a pretty horrible use of std::set. Use a vector, allocate once up front, and later on sort & filter. Then later, both iterating the keys, or using std::lower_bound, would be an improvement over the set.  Something along these lines
```c++
    unsigned int numFrameSamples = argData.numberOfFlagUses(kFrameSampleFlag);
    std::vector<double> frameSamples(numFrameSamples);

    for (unsigned int i = 0; i < numFrameSamples; ++i) {
        MArgList tmpArgList;
        argData.getFlagArgumentList(kFrameSampleFlag, i, tmpArgList);
        frameSamples[i] = tmpArgList.asDouble(0);
    }
    std::sort( frameSamples.begin(), frameSamples.end() ) ;
    frameSamples.erase( std::unique( frameSamples.begin(), frameSamples.end() ), frameSamples.end() ) ;
```

Reply:
Hey @robthebloke , 

First apologies for leaving this review for so long.  All of this is copied code from the Pixar plugin down into the core and this draft review was mostly put here to give people the heads-up that it was getting moved.  We had this in-progress, but we knew we were going to pause work on a bit and wouldn’t be merged until at least some automated tests were added.  I didn’t even realized that you’d commented here until just recently but have made most of the changes you pointed out (and will push in a bit).

This one though…  The interface to UsdMayaWriteUtil::GetTimeSamples requires a std::set<double> so we do need that.  And if I understand how the “frame samples” flag is typically used, I think we would only normally have a couple values added here, so I don’t think pre-allocating a vector + sorting + filter is really required.